demo/src/core/spinner-component.js

```diff
@@ -0,0 +1,88 @@
+import { parseHtml } from './html-utils';
+
+let SPINNER_ID = 0;
+
+/**
+ * A simple spinner component that can be attached to a DOM element.
+ * @class
+ */
+class SpinnerComponent {
+  /**
+   * A private property to track the visibility state of the spinner.
+   *
+   * @private
+   * @member {boolean}
+   */
+  #hidden = true;
+
+  constructor() {
+    /**
+     * The unique identifier for the spinner instance and DOM element.
+     *
+     * @member {string}
+     */
+    this.id = `spinner-${SPINNER_ID += 1}`;
+  }
+
+  /**
+   * Attach the spinner to a specified DOM element.
+   *
+   * @param {string} selector - The CSS selector for the target DOM element.
+   *
+   * @returns {SpinnerComponent} - The current SpinnerComponent instance for chaining.
+   *
+   * @throws Error Will throw an error if the spinner is already attached.
+   */
+  attach(selector) {
+    if (document.querySelector(`#${this.id}`)) {
+      throw new Error(`Spinner ${this.id} already exists, create a new spinner`);
+    }
+
+    document.querySelector(selector).appendChild(parseHtml(`
+      <div id="${this.id}" class="spinner-container hidden">
+        <div class="spinner"></div>
+      </div>
+    `)[0]);
+
+    document.querySelector(`#${this.id}`).addEventListener('animationend', (event) => {
+      event.target.classList.remove('slide-fade');
+    });
+
+    return this;
+  }
+
+  /**
+   * Show the spinner.
+   */
+  show() {
+    if (this.hidden) {
+      document.querySelector(`#${this.id}`).classList.remove('hidden');
+      document.querySelector(`#${this.id}`).classList.add('slide-fade');
```

Comment:
Is it fine to query the same selector twice?
```suggestion
      let spinnerClassList = document.querySelector(`#${this.id}`).classList;
      spinnerClassList.remove('hidden');
      spinnerClassList.add('slide-fade');
```

The same applies in `hide()` below.

Reply:
Spot on ! I've refactor this module to reduce the number calls to the `querySelector`. I've applied the same principle to the search page since it's part of this feature.